src/uct/base/uct_iface.h

```diff
@@ -396,6 +396,59 @@ typedef struct uct_iface_local_addr_ns {
     }
 
 
+#define uct_tl_register(_name) \
+    uct_tl_##_name##_ctor()
+
+
+#define uct_tl_unregister(_name) \
+    uct_tl_##_name##_dtor()
+
+/**
+ * Declare transport registration routines
+ *
+ * @param [in] _name Transport name used to generate ctors/dtors.
+ * @param [in] ...   Routine attributes (static)
+ */
+#define UCT_TL_REGISTER_DECL(_name, ...) \
+    __VA_ARGS__ void uct_tl_##_name##_ctor(void); \
+    __VA_ARGS__ void uct_tl_##_name##_dtor(void);
+
+/**
+ * Transport registration routines
+ *
+ * @param _component      Component to add the transport to
+ * @param _name           Name of the transport (should be a token, not a string)
+ * @param _query_devices  Function to query the list of available devices
+ * @param _iface_class    Struct type defining the uct_iface class
+ * @param _cfg_prefix     Prefix for configuration variables
+ * @param _cfg_table      Transport configuration table
+ * @param _cfg_struct     Struct type defining transport configuration
+ */
+#define UCT_TL_REGISTER_DEF(_component, _name, _query_devices, _iface_class, \
+                            _cfg_prefix, _cfg_table, _cfg_struct) \
+    \
+    static uct_tl_t uct_##_name##_tl = { \
+        .name               = #_name, \
+        .query_devices      = _query_devices, \
+        .iface_open         = UCS_CLASS_NEW_FUNC_NAME(_iface_class), \
+        .config = { \
+            .name           = #_name" transport", \
+            .prefix         = _cfg_prefix, \
+            .table          = _cfg_table, \
+            .size           = sizeof(_cfg_struct), \
+         } \
+    }; \
+    void uct_tl_##_name##_ctor(void) \
+    { \
+        ucs_list_add_tail(&ucs_config_global_list, &(uct_##_name##_tl).config.list); \
+        ucs_list_add_tail(&(_component)->tl_list, &(uct_##_name##_tl).list); \
+    } \
+    void uct_tl_##_name##_dtor(void) \
+    { \
+        ucs_list_del(&(uct_##_name##_tl).config.list); \
```

Comment:
need to remove from ucs_config_global_list

Reply:
you mean that because other modules do not remove entries, other modules would segfualt?

Reply:
right. it depends from module unload sequence